packages/flutter/lib/src/material/date_picker.dart

```diff
@@ -1425,18 +1436,23 @@ class _DateRangePickerDialogState extends State<DateRangePickerDialog> with Rest
 
         case DatePickerEntryMode.input:
           // Validate the range dates
+
+          // If start date after end date, then just use the start date
+          if (_selectedStart.value != null && _selectedEnd.value != null && _selectedStart.value!.isAfter(_selectedEnd.value!)) {
+            _selectedEnd.value = null;
+          }
+          // If start date is outside expected start and end date range, or is not selectable, then clear both start and end dates
           if (_selectedStart.value != null &&
-              (_selectedStart.value!.isBefore(widget.firstDate) || _selectedStart.value!.isAfter(widget.lastDate))) {
+              (_selectedStart.value!.isBefore(widget.firstDate) || _selectedStart.value!.isAfter(widget.lastDate) ||
+                  widget.selectableDayPredicate != null && !widget.selectableDayPredicate!(_selectedStart.value!, _selectedStart.value, _selectedEnd.value))) {
             _selectedStart.value = null;
             // With no valid start date, having an end date makes no sense for the UI.
             _selectedEnd.value = null;
           }
+          // If end date is outside expected start and end date range, or is not selectable, then clear the end date
           if (_selectedEnd.value != null &&
-              (_selectedEnd.value!.isBefore(widget.firstDate) || _selectedEnd.value!.isAfter(widget.lastDate))) {
-            _selectedEnd.value = null;
-          }
-          // If invalid range (start after end), then just use the start date
-          if (_selectedStart.value != null && _selectedEnd.value != null && _selectedStart.value!.isAfter(_selectedEnd.value!)) {
+              (_selectedEnd.value!.isBefore(widget.firstDate) || _selectedEnd.value!.isAfter(widget.lastDate) ||
+                  widget.selectableDayPredicate != null && !widget.selectableDayPredicate!(_selectedEnd.value!, _selectedStart.value, _selectedEnd.value))) {
             _selectedEnd.value = null;
           }
           _entryMode.value = DatePickerEntryMode.calendar;
```

Comment:
I added the selectableDayPredicate check when switching from input to calendar mode, now it "unselect" the non-selectable dates. The corresponding 2 tests have also been added, named `Non-selectable start date` and `Non-selectable end date`.

I moved the "start cannot be after the end" block to the beginning to ensure the start & end dates passed to `selectableDayPredicate` are always in the good order.